### DISCUSSION TITLE
BlazorServer : ClaimsPrincipal issue

### DISCUSSION THREAD

**Discussion by sylum** [MarimerLLC/csla#2084] [Questions] [answered]
Hi Folk,

I'm facing an issue with the BlazorServer application. 
Actually, the application works fine when running on IISExpress but  failed on IIS.
The issue came after been authenticated  (manager/manager) on the first call to AppServer (when fetching the data). 
I have got the following error message : 
**"Unable to cast object of type 'Csla.Security.UnauthenticatedPrincipal' to type 'System.Security.Claims.ClaimsPrincipal"**

```
Csla.Server.DataPortalException: DataPortal.Fetch a échoué sur le serveur
at Csla.AspNetCore.ApplicationContextManager.SetUser(IPrincipal principal)
at Csla.Server.DataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
at Csla.Server.DataPortal.NewDataPortalException(String message, Exception innerException, Object businessObject)
at Csla.Server.DataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
at Csla.Server.Hosts.HttpPortal.Fetch(CriteriaRequest request)
at Csla.Server.Hosts.HttpPortalController.CallPortal(String operation, Object request)
at Csla.Server.Hosts.HttpPortalController.InvokePortal(String operation, Stream requestStream, Stream responseStream)
System.InvalidCastException: Unable to cast object of type 'Csla.Security.UnauthenticatedPrincipal' to type 'System.Security.Claims.ClaimsPrincipal'.
at Csla.AspNetCore.ApplicationContextManager.SetUser(IPrincipal principal)
at Csla.Server.DataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
```

In debug, the section below works well ans the cookie is well created

```
 var validator = await Csla.DataPortal.FetchAsync<CredentialValidator>(credentials);
      var principal = validator.GetPrincipal();

      Csla.ApplicationContext.User = principal;

      if (principal.Identity.IsAuthenticated)
      {
        AuthenticationProperties authProperties = new AuthenticationProperties();
        await HttpContext.SignInAsync(
          CookieAuthenticationDefaults.AuthenticationScheme,
          principal,
          authProperties);

        string returnUrl = Url.Content("~/");
        return LocalRedirect(returnUrl);
      }
```

Does someone has already encounter this issue ?

Thanks a lot for you feedback in advance,

Best regards,
Cedric

**Version and Platform**
CSLA : 5.3.1
BLAZOR SERVER
VS2019
WINDOW10

**Comment by rockfordlhotka** [comment]
aspnetcore and Blazor do not work with anything other than `ClaimsPrincipal` and `ClaimsIdentity`.

In fact, in CSLA 6 I am removing all the old principal/identity types from the framework, as they've become effectively useless due to Microsoft's changes.

In CSLA 5.4.0 (I think) the default unauthenticated principal was changed from `UnauthenticatedPrincipal` to an unauthenticated `ClaimsPrincipal`. You can do that by hand, as your app starts up just make sure to run this code

```c#
  Csla.ApplicationContext.User = new ClaimsPrincipal();
```

**Comment by sylum** [comment, marked as answer]
Hello Rocky,

Thx for your answer, I have followed your advice and added the code below into the "Startups.cs" without success.

```
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
...
            app.UseCsla(c => c
              .DataPortal()
              .DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy), Configuration.GetSection("AppSettings:DataPortalUrl").Value));
            Csla.ApplicationContext.User = new ClaimsPrincipal();
}
```
Next, I have upgraded the solution with the latest CSLA version (4.5.1) but i got same error with the ServerSide Blazor.
By the way the ClientSide Blazor  works perfectly with both CSLA version(using multi-head solution)

Below package used by the project  ServerSide Blazor :
```
	<ItemGroup>
		<PackageReference Include="Csla.AspNetCore" Version="5.4.1" />
		<PackageReference Include="Csla.Blazor" Version="5.4.1" />
		<PackageReference Include="Microsoft.AspNetCore.Blazor.HttpClient" Version="3.2.0-preview3.20168.3" />
		<PackageReference Include="Microsoft.AspNetCore.Components.Authorization" Version="3.1.9" />
		<PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="3.1.8" />
		<PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.4" />
	</ItemGroup>
```
Finally, I have double checked if all IIS components et SDK/Runtime Core are well installed.

I must missing something but I dont give up.

Thanks for your help,

Cedric

**Reply by rockfordlhotka** [reply]
I see that you are using the data portal via HttpProxy, and it appears that the exception is occurring on the appserver.

The appserver should be getting the current principal from the client: it is passed through the data portal from client to server.

I would set a breakpoint on your client-side `DataPortal.FetchAsync` (or equivalent) method and use the Immediate window to verify that the `ApplicationContext.User` property is correct (it should be a `ClaimsPrincipal`).

**Reply by sylum** [reply]
Hey it was quick.
I have cloned the solution into another PC and now all working fine with both Blazor app (Server Side & Client Side)
So, when I will get the hand on the rebel PC (at the office), I will continu the investigation with your advice to anderstand the issue.
Thx a lot, appreciate your help.

**Comment by MarkOverstreet** [comment]
Quick question that is possibly related to this...

I have a small Blazor server project and I tried using the same code as in the Using CLSA 5: Blazor and WebAssembly book but I'm having an issue in that the Csla.ApplicationContext.User is not being set automatically as indicated on Page 137.  Here is the testing code ....

```
        public async Task<IActionResult> OnPostAsync()
        {
            if (!ModelState.IsValid)
            {
                ErrorText = "Form has validation errors.";
                return Page();
            }

            if (!string.IsNullOrWhiteSpace(Username) &&
                Users.TryGetValue(Username.ToLower(), out string pw))
            {
                if (pw == Password)
                {
                    var identity = new ClaimsIdentity("password");
                    identity.AddClaim(new Claim(ClaimTypes.Name, Username.ToLower()));
                    //identity.AddClaim(new Claim(ClaimTypes.Role, "admin"));
                    identity.AddClaim(new Claim(ClaimTypes.Role, pw));
                    var principal = new ClaimsPrincipal(identity);
                    var authProperties = new AuthenticationProperties();
                    await HttpContext.SignInAsync(
                      CookieAuthenticationDefaults.AuthenticationScheme,
                      principal,
                      authProperties);

                    return LocalRedirect(Url.Content("~/dashboard"));
                }
            }
            ErrorText = "Invalid credentials";
            return Page();
        }
    }
```

I tried to add this `Csla.ApplicationContext.User= principal`  after this statement `var principal = new ClaimsPrincipal(identity);`

But still I have Csla.Security.UnauthenticatedPrincipal when I try to use it on one of my pages?  Any thoughts?

I tried looking at your example projects (ex:  BlazorAuthServer) but I couldn't find good example that demonstrated this.
